### ISSUE TITLE
Speculative verification sorts the whole vocabulary per row when top_k <= 0, making the drafter a 2x net loss

### ISSUE THREAD

**Issue by dlsniper** [ardanlabs/kronk#934]
## Summary

`top_k <= 0` makes every classic speculative verification row sort the entire vocabulary. With the separate-GGUF drafter configured this costs far more than the drafter saves: 88 tok/s with `top_k: 0` vs 162 tok/s with `top_k: 20`, and 196 tok/s with the drafter removed entirely.

## Reachable input

`top_k` is stored unvalidated (`sdk/kronk/model/params.go:612-617`). The only normalization is `adjustParams` at `params.go:761`, guarded by `!requested("top_k") && !m.paramsResolved && p.TopK == 0` — and `paramsResolved` is always `true` for a loaded model (`sdk/kronk/model/model.go:399`), so an explicitly requested `top_k` is never rewritten. This is deliberate: `TestParseParamsSamplingSentinels` (`sdk/kronk/model/chat_test.go:2330-2339`) pins `top_k: 0` and `top_k: -1` as "disables filtering" sentinels. `/v1/chat/completions` decodes the request body straight into `model.D` (`cmd/server/app/domain/chatapp/chatapp.go:30-57`), so the value arrives verbatim over the wire.

## Path

- `sdk/kronk/model/speculation_runtime.go:268-270` — the classic verify `Target` callback calls `applySamplerFilters(..., s.job.params.TopK, ...)`. `classic.Verify` invokes it once per drafted candidate plus once for the bonus row (`sdk/kronk/model/internal/speculation/classic/verify.go:48,86`), i.e. up to `nDraft+1` times per speculative round.
- `sdk/kronk/model/logprobs.go:134-137` — `selectK := n; if topK > 0 { selectK = int(topK) }`. With `topK <= 0`, `selectK` is the full vocab.
- `sdk/kronk/model/logprobs.go:152-196` — the min-heap therefore admits every token and `sort.Slice` (`:194`) sorts all `n` indices. For Qwen3 that is 151,936 elements sorted up to five times per round.

Only the classic controller reaches this (`speculation_runtime.go:19-20`). Greedy requests (`temperature == 0`) take the `GreedyToken` branch at `speculation_runtime.go:264-267` and are unaffected, as is non-speculative generation.

## Consequence

macOS/Metal, `testlib.CfgClassicDraftChat()` (Qwen3-1.7B-Q4_K_M target + Qwen3-0.6B-Q8_0 draft), `temperature: 0.9`, `seed: 11`, `max_tokens: 300`, same prompt, alternated, two runs each:

```
top_k=20   1.86s  162 tok/s   draft=110  accepted=39  rate=0.35
top_k=0    3.42s   88 tok/s   draft=80   accepted=26  rate=0.33
top_k=-1   3.40s   88 tok/s   draft=80   accepted=26  rate=0.33
```

Same config with `PtrDraftModel = nil`: `top_k=20` → 209 tok/s, `top_k=0` → 196 tok/s. So disabling top-k costs 6% without a drafter and 46% with one, and under `top_k: 0` configuring the drafter is a 2.2x net loss versus not configuring it.

Acceptance rate is essentially unchanged (0.35 vs 0.33), so this is not proposal quality — it is the sort. CPU profile of the `top_k: 0` run:

```
      flat  flat%   sum%        cum   cum%
     1.95s 30.28% 30.28%      2.83s 43.94%  sort.partition_func
     0.96s 14.91% 84.94%      0.96s 14.91%  model.applySamplerFilters.func1
     0.13s  2.02% 95.03%      3.60s 55.90%  model.applySamplerFilters
         0     0% 98.60%      3.60s 55.90%  model.(*batchEngine).ClassicVerifyInput.func1
```

The identical run at `top_k: 20` does not show `applySamplerFilters` in the profile at all (1.57s total samples vs 6.44s).

Secondary, same trigger and much smaller (~2% of the profile): `buildDraftSampler` (`sdk/kronk/model/model.go:1074`) passes `TopK` through unguarded too, so with `top_k <= 0` the draft chain's sampled-candidate readback returns the full vocabulary and `llama.DraftGenerate` copies all of it into `s.draftCandDistBuf` per drafted token — buffers pre-sized at `cap 128` in `sdk/kronk/model/batchgen_speculative.go:192-198`.

## Repro

Drop into `sdk/kronk/tests/draft` and run with `-run TestTopKZeroDraftCost -v`:

```go
func TestTopKZeroDraftCost(t *testing.T) {
	testlib.WithModel(t, testlib.CfgClassicDraftChat(), func(t *testing.T, krn *kronk.Kronk) {
		for _, topK := range []int{20, 0, 20, 0} {
			d := model.D{
				"messages": []model.D{{"role": "user",
					"content": "Write a short surreal story about a lighthouse keeper who collects thunderstorms."}},
				"max_tokens": 300, "temperature": 0.9, "seed": 11,
				"enable_thinking": false, "top_k": topK,
			}
			ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
			start := time.Now()
			resp, err := krn.Chat(ctx, d)
			cancel()
			if err != nil {
				t.Fatalf("top_k=%d: %v", topK, err)
			}
			t.Logf("top_k=%-3d elapsed=%-9s tps=%6.2f draft=%d accepted=%d",
				topK, time.Since(start).Round(time.Millisecond),
				resp.Usage.TokensPerSecond, resp.Usage.DraftTokens, resp.Usage.DraftAcceptedTokens)
		}
	})
}
```

`applySamplerFilters` currently has no test for `topK <= 0`; `TestApplySamplerFiltersSuppressesBeforeTopK` (`sdk/kronk/model/logprobs_test.go:140`) passes `topK = 1`.

Bounding `selectK` when top-k is disabled — or partial-sorting instead of a full `sort.Slice` — is the obvious guard.

**Comment by ardan-bkennedy** [comment]
Thanks for the profile and the clear repro, Florin. The proposed direction for addressing this over the next few days is:

1. Preserve `top_k <= 0` as “top-k disabled.” Do not impose an arbitrary hidden candidate cap, since that would change the requested distribution.
2. Keep the optimization in the classic verification distribution path rather than reuse the live `toSampler()` sampler. That sampler owns RNG and history-dependent request state and should not be advanced for speculative rows that may be rejected.
3. Optimize `applySamplerFilters` for the no-top-k case while preserving llama.cpp semantics. The likely approach is to mirror b10819's native filtering strategy: unsorted softmax, adaptive partial selection for top-p (starting with a small candidate set and expanding only when necessary), and a linear min-p path where the sampler order allows it. A dedicated native filter-only chain remains an alternative if exact Go parity proves too brittle, but it does not seem necessary as the first implementation.
4. Add differential tests against the native llama.cpp sampler chain before changing the algorithm. A related `top_p: 0` parity discrepancy surfaced during review and is tracked separately in #936.
5. Benchmark a Qwen-sized synthetic vocabulary and rerun the reported `top_k` values (`20`, `0`, `-1`). Then reprofile before deciding whether the smaller full-vocabulary draft-candidate copy deserves a separate optimization.

The important constraint is that the optimized reconstruction must produce the same candidate support and normalized probabilities as llama.cpp; performance should not come from silently turning disabled top-k into a finite top-k.

Florin, does this direction match what you had in mind? In particular, any preference between keeping an exact Go reconstruction and using a separate native filter-only sampler chain for verification?

**Comment by dlsniper** [comment]
llama.cpp refs below are from `.extras/llama.cpp` @ `a94d563ed` (`b10414-9`). Plan cites b10819; `README.md:118` pins b10842. Reconfirm on the pinned build.

**P3 corrections**

- Drop the min-p linear path. Native takes it only when `!sorted`; top-p sets `sorted` for `p < 1`. Order is top-k→top-p→min-p (`params.go:911,915,919`; `model.go:1066-1068`), `DefTopP=0.95` (`params.go:105`), `DefMinP=0` (`:69`). Unreachable in the repro, contributes 0.
- Mirror is unbounded. Native expansion goes `k=256` → `k=size`; partial-sort at `npartial==size` is a full sort. `top_p`→1 on flat rows keeps today's cost. State the terminal strategy.
- Parity goal unachievable as written. `applySamplerFilters` models 4 samplers (`logprobs.go:117-128`); live chain adds penalties/DRY/XTC/adaptive-p (`params.go:896,906,924,930`), none gated for speculation; non-greedy emits from the reconstruction (`speculation_runtime.go:271`→`classic/verify.go:109-113`). Scope the reference chain to those 4, file the rest.

**P4 corrections**

- Covers `p_target` only. `q_draft` comes from a backend/graph sampler (`batchgen_speculative.go:169-181`, yzma `draft.go:98-125`) with different semantics (backend top-p masks, ignores `min_keep`). `llama_sampler_apply` won't test it.
- No probability readback. `SamplerInitTempExt` rescales logits, doesn't recompute `p`; no `llama_sampler_init_softmax` exported. Pick now: append `SamplerInitDist`, or compare id-set + post-chain logits and softmax in Go.
- Tie/precision policy required. Both sorts unstable; native cumsum `float` vs kronk `float64` (`logprobs.go:194,210,223`). Cutoff can differ by one over 151936 terms.
- CI landmine: macOS race job is un-shimmed on the invariant that `-short` skips Metal in `sdk/kronk/model` (`.github/workflows/macos.yml:134-136`). No `TestMain`, no tags, no `Short()` gate there.

**Sequencing**

- #936 is not separable. Native top-p means replacing `topP > 0 && topP < 1.0` (`logprobs.go:222`) with a `p >= 1` no-op; that is the #936 fix. #936 first, or ship P4's `top_p: 0` case expected-fail.

**Land first (unconditional, no #936 dependency)**

- `topK<=0 && topP>=1 && minP<=0`: skip ordering entirely. Nothing consumes it (`logprobs.go:222,234`); temp softmax is order-independent (`:255-267`). One `if`.
- `topK<=0`: skip the heap. `selectK==n` sifts every element (`logprobs.go:152-166`) into a heap re-sorted at `:194`. Fill `indices` `0..n-1`.
- Scope `clear(probs)` (`logprobs.go:248`) to the returned survivor set: currently `nVocab` floats × (nDraft+1) per round at every `top_k` (`model.go:1053`).

**Verify before P5**

- `top_k<=0 && top_p>=1` may accept 0 tokens/round: all truncation samplers no-op → no candidates assigned → `cnt==0` (`draft.go:105`) → empty `outDists[i]` → `classic/verify.go:57` bails at i=0, still paying the full sort. `top_p: 1` is the OpenAI default.

**P5 corrections**

- Zero benchmarks exist in `sdk/`, no `-bench` in CI. `BenchmarkApplySamplerFilters`, 151936 synthetic vocab, top_k × top_p × min_p matrix, committed as the gate. Metal tok/s won't reproduce in CI.
- `top_k: -1` ≡ `0` (`logprobs.go:135`), same native no-op. Drop it.
- Draft-candidate copy isn't the only O(V) cost at `topK<=0`: `classic/verify.go:132-151`, `:153-165` (no early exit), `:187-195` O(V·nDraft) via `lookupProbability` (`:123-130`).
- `s.draftCandDistBuf` grows to full vocab and `slot.reset` never clears it (`batchgen_speculative.go:192-198`), leaving ~6 MB/slot retained.

**Severity**

- Not request-only: `resolveSamplingDefaults` substitutes `DefTopK` only at `p.TopK == 0` (`params.go:826-828`); `WithDefaultParams` with `TopK: -1` bypasses it.

**Your question**

- RNG/history objection applies to the live chain only: top-k/top-p/min-p/temp are stateless. Real cost of a native filter-only chain: O(nVocab) `[]TokenData` (~1.8 MB) + FFI per row. Exposed (`sampling.go:627`, `llama.go:280-292`), precedent at `grammar.go:399-435`. Benchmark it against an O(n) Go path, not today's O(n log n).

**Comment by ardan-bkennedy** [comment]
Fixed with PR https://github.com/ardanlabs/kronk/pull/945
